scri/bms_transformations.py

```diff
@@ -0,0 +1,829 @@
+    
+def transform_Y_LM(Y_LM, Lorentz, ell_max=None):
+    """Apply a Lorentz transformation to a spherical harmonic.
+
+    Parameters
+    ----------
+    f_LM: ndarray, dtype=complex
+        Ylms to be transformed.
+    Lorentz: Lorentz_transformation
+        Lorentz transformation to be used to transform the function.
+    ell_max: int
+        Maximum ell to use when expressing functions via coordinates on the two-sphere.
+        Defaults to the ell_max of the Lorentz transformation.
+    """
+    if ell_max is None:
+        ell_max = Lorentz.ell_max
+    
+    spin_matrix = Lorentz_to_spin_matrix(Lorentz)
+
+    from_zprime_to_z = []
+    from_SH_expansion = []
+    for N1 in range(Y_LM[0] + 1):
+        for N2 in range(Y_LM[0] + 1):
+            factor = 0
+            for M1 in range(Y_LM[0] + 1):
+                for M2 in range(Y_LM[0] + 1):
+                    factor += A_matrix(Y_LM[0], M1, M2, N1, N2, spin_matrix)\
+                        *B_matrix(Y_LM[0], Y_LM[0], Y_LM[1], M1, M2)
+            from_zprime_to_z.append(factor)
+
+            factors = []
+            for L in range(0, Y_LM[0] + 1):
+                for M in range(-L, L + 1):
+                    factors.append(B_matrix(Y_LM[0], L, M, N1, N2))
+            from_SH_expansion.append(factors)
+            
+    n_theta = 2 * ell_max + 1; n_phi = n_theta;
+            
+    f_LM = np.matmul(np.linalg.inv(np.array(from_SH_expansion)),np.array(from_zprime_to_z))
+    f_LM_padded = np.pad(f_LM, (0,(ell_max + 1)**2 - f_LM.size))
+    f_LM_S2 = spinsfast.salm2map(f_LM_padded, 0, ell_max, n_theta, n_phi)
+    
+    K_S2 = compute_conformal_factor(Lorentz, n_theta, n_phi)
+
+    result_S2 = K_S2**(Y_LM[0]) * f_LM_S2
+    
+    return spinsfast.map2salm(result_S2, 0, ell_max)
+
+def transform_function(f_LM, Lorentz, ell_max=None, tol=1e-15):
+    """Apply a Lorentz transformation to a spin-weight 0 funtion, where the function
+    is provided via it's spherical harmonic modes, the Ylms.
+
+    Parameters
+    ----------
+    f_LM: ndarray, dtype=complex
+        Ylms to be transformed.
+    Lorentz: Lorentz_transformation
+        Lorentz transformation to be used to transform the function.
+    ell_max: int
+        Maximum ell to use when expressing functions via coordinates on the two-sphere.
+        Defaults to the ell_max of the Lorentz transformation.
+    tol: float
+        Tolerance to use when rewriting the supertranslation's transformed Ylms
+        in terms of the untransformed Ylms. Defaults to 1e-15.
+    """
+    if ell_max is None:
+        ell_max = Lorentz.ell_max
+
+    f_LM_prime = np.zeros((ell_max + 1)**2, dtype=complex)
+    for L in range(0, int(np.sqrt(f_LM.size) - 1) + 1):
+        for M in range(-L, L+1):
+            if abs(f_LM[sf.LM_index(L, M, 0)]) < tol:
+                continue
+            f_LM_prime += f_LM[sf.LM_index(L, M, 0)] * transform_Y_LM((L, M), Lorentz, ell_max)
+
+    return f_LM_prime
```

Comment:
I feel like this function is doing way more work than it needs to... every Y_{\ell m} is being separately transformed on a grid and the transformed data converted back from position space to harmonic space... and then summing over all \ell m. But I feel that it must be possible to do this with only one transformation, rather than (\ell+1)^2 transformations.